File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
# 2156
import sys
input = sys.stdin.readline
n = int(input())
dp = [0] * 10000
arr = []

for _ in range(n):
    arr.append(int(input()))

if n == 1:
    print(arr[0])
elif n == 2:
    print(arr[0] + arr[1])
elif n == 3:
    print(max(arr[1] + arr[2], arr[0] + arr[2], arr[0] + arr[1]))
else:
    dp[0] = arr[0]    
    dp[1] = arr[0] + arr[1]
    dp[2] = max(arr[1] + arr[2], arr[0] + arr[2], arr[0] + arr[1])
    for i in range(3, n):
        dp[i] = max(dp[i-2] + arr[i], dp[i-3] + arr[i-1] + arr[i], dp[i-1])
    print(dp[n-1])

File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n];
        int[][] dp = new int[n][2];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        dp[0][1] = arr[0];
        if (n == 1) {
            System.out.println(arr[0]);
        } else {
            dp[1][0] = arr[0];
            dp[1][1] = arr[0] + arr[1];
            for (int i = 2; i < n; i++) {
                dp[i][0] = Math.max(dp[i-1][0], dp[i-1][1]);
                dp[i][1] = Math.max(dp[i-2][0] + arr[i-1] + arr[i], dp[i-2][1] + arr[i]);
            }
            System.out.println(Math.max(dp[n-1][0], dp[n-1][1]));
        }

    }
}
